extensions/standard-processors/utils/JoltUtils.cpp

```diff
@@ -0,0 +1,1120 @@
+    // literal is matched
+    Context new_ctx = ctx.extend({name}, &member);
+    process(std::get<2>(literals.at(it->second)), new_ctx, member, output);
+    return true;
+  }
+  for (auto& templ : templates) {
+    if (templ.first.eval(ctx) == name) {
+      Context new_ctx = ctx.extend({name}, &member);
+      process(templ.second, new_ctx, member, output);
+      return true;
+    }
+  }
+  for (auto& reg : regexes) {
+    if (auto matches = reg.first.match(name)) {
+      Context new_ctx = ctx.extend(matches.value(), &member);
+      process(reg.second, new_ctx, member, output);
+      return true;
+    }
+  }
+  return false;
+}
+
+void Spec::Pattern::process(const Context& ctx, const rapidjson::Value &input, rapidjson::Document &output) const {
+  auto on_exit = ctx.log([&] (auto logger) {
+    logger->log_trace("Processing node at {}", ctx.path());
+  }, [&] (auto logger) {
+    logger->log_trace("Finished processing node at {}", ctx.path());
+  });
+  for (auto& [val_ref, dest] : values) {
+    auto& [idx, path] = val_ref;
+    auto* target = ctx.find(idx);
+    if (!target) {
+      throw Exception(GENERAL_EXCEPTION, fmt::format("Could not find parent node at offset {} for {}", idx, ctx.path()));
+    }
+    if (!target->node) {
+      return;
+    }
+    if (auto value = resolvePath(ctx, target->path(), *target->node, path)) {
+      Context sub_ctx = ctx.extend(ctx.matches, ctx.node);
+      process(dest, sub_ctx, value.value(), output);
+    } else {
+      ctx.log([&, path_ptr = &path] (auto logger) {
+        auto path_str = toString(ctx, *path_ptr);
+        logger->log_trace("Failed to resolve value path {} (evaled as {}) at {} (triggered from {}): {}", path_str.first, path_str.second, target->path(), ctx.path(), value.error());
+      }, [&] (auto) {});
+      // pass, non-existent member is not an error
+    }
+  }
+  for (auto& [key, dest] : keys) {
+    auto key_str = ctx.find(key.first)->matches.at(key.second);
+    putValue(ctx.extend({key_str}, nullptr), dest, rapidjson::Value{key_str.data(), gsl::narrow<rapidjson::SizeType>(key_str.size()), output.GetAllocator()}, output);
+  }
+  for (auto& [value, dest] : defaults) {
+    Context sub_ctx = ctx.extend( {value}, nullptr);
+    putValue(sub_ctx, dest, rapidjson::Value{value.data(), gsl::narrow<rapidjson::SizeType>(value.size()), output.GetAllocator()}, output);
+  }
+  if (input.IsArray()) {
+    Context sub_ctx = ctx;
+    for (auto& [key, numeric_key, value] : literals) {
+      if (numeric_key && numeric_key.value() < input.GetArray().Size()) {
+        if (processMember(sub_ctx, key, input[numeric_key.value()], output)) {
+          ++sub_ctx.match_count;
+        }
+      }
+    }
+    for (rapidjson::SizeType  i = 0; i < input.GetArray().Size(); ++i) {
+      if (literal_indices.contains(std::to_string(i))) {
+        continue;
+      }
+      if (processMember(sub_ctx, std::to_string(i), input[i], output)) {
+        ++sub_ctx.match_count;
+      }
+    }
```

Comment:
Could you extract this into a processArray function for increased readability?

Reply:
done